### PR TITLE
Feature/49

### PR DESCRIPTION
**Pull request by mpkato**
#49

### CODE REVIEW

**Comment by mpkato** [comment]
Sorry, I have just found another bug. Let me update the code before merging.

**Comment by mpkato** [comment]
Two things need to be addressed:
- R_used is broken by the sampling
- Need careful treatment for log-exp transformation

File: interleaving/probabilistic.py [hunk 1]
```diff
@@ -178,31 +172,53 @@ def _compute_scores(cls, ranking, clicks, tau=3.0, n=10**4):
...
+                    continue
+
+                A, A_prime = A_prime, []
+                is_pass = np.random.rand(len(A), len(R_non_zero)) <= threshold
+                for i, (o, p, a) in enumerate(A):
+                    # Skip some assignments with certain probability
+                    R_used = [R_non_zero[k] for k in range(len(R_non_zero))
+                        if is_pass[i][k]]
+                    for j, R_j in R_used:
+                        p_prime = p + np.log(P[j])
```
Comment (tmanabe) [inline interleaving/probabilistic.py]: I have no idea why `/ 2` is removed.

File: tests/test_balanced.py [hunk 2]
```diff
@@ -3,7 +3,6 @@
 import json
 import numpy as np
 import pytest
-np.random.seed(0)
```
Comment (tmanabe) [inline tests/test_balanced.py]: Why did you remove this line?

File: interleaving/probabilistic.py (hunk 1 above)
Comment (mpkato) [reply interleaving/probabilistic.py]: I assumed that the original version is already wrong.
I think 1/2 comes from Probabilistic Interleaving, where P(a|l, q) = P(l|a, q)P(a|q)/P(l|q) and P(a|q) = 1/(the number of rankers). P(a|q) = 1/2 for PI. Thus, 1/2 may represent P(a|q).

Although this part must be updated for PM, i.e. `1/len(R)`,
this term does not affect the final outcome and can be removed as follows:

`P(a|l, q) = P(l|a, q)P(a|q)/P(l|q) = P(l|a, q)P(a|q)/sum_a P(l|a, q)P(a|q) = P(l|a, q)/sum_a P(l|a, q)`

where P(a|q) = 1/(the number of rankers) and does not depend on `a`.

File: tests/test_balanced.py (hunk 2 above)
Comment (mpkato) [reply tests/test_balanced.py]: Please also look at the parent class TestMethods.fix_seed in tests/test_methods.py .
All the test classes for interleaving inherits this class and call TestMethods.fix_seed every time before a test method is called.
By doing this, we can fix the seed for all the test methods, which is independent from the execution order of test methods.

File: interleaving/probabilistic.py (hunk 1 above)
Comment (tmanabe) [reply interleaving/probabilistic.py]: I understand. Thank you!

File: tests/test_balanced.py (hunk 2 above)
Comment (tmanabe) [reply tests/test_balanced.py]: I overlooked that. Thank you!

**Comment by tmanabe** [comment]
Thank you again for your useful PR.
You can merge this PR in your convenience.

**Comment by mpkato** [comment]
Thanks, @tmanabe
